packages/firestore/lite/src/api/reference.ts

```diff
@@ -49,3 +55,222 @@ export class DocumentReference<T = firestore.DocumentData>
+  collectionPath: string
+): CollectionReference<firestore.DocumentData>;
+export function collection(
+  parent: firestore.FirebaseFirestore | firestore.DocumentReference<unknown>,
+  relativePath: string
+): CollectionReference<firestore.DocumentData> {
+  if (relativePath.length === 0) {
+    throw new FirestoreError(
+      Code.INVALID_ARGUMENT,
+      `Invalid path (${relativePath}). Empty paths are not supported.`
+    );
+  }
+
+  const path = ResourcePath.fromString(relativePath);
+  if (parent instanceof Firestore) {
+    if (DocumentKey.isDocumentKey(path)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${path}). Path points to a document.`
+      );
+    }
+    return new CollectionReference(parent, path);
+  } else {
+    const doc = hardCast(parent, DocumentReference);
+    const absolutePath = doc._key.path.child(path);
+    if (DocumentKey.isDocumentKey(absolutePath)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${absolutePath}). Path points to a document.`
+      );
+    }
+    return new CollectionReference(doc.firestore, absolutePath);
+  }
+}
+
+export function doc(
+  firestore: FirebaseFirestore,
+  documentPath: string
+): DocumentReference<firestore.DocumentData>;
+export function doc<T>(
+  reference: CollectionReference<T>,
+  documentPath?: string
+): DocumentReference<T>;
+export function doc<T>(
+  parent: firestore.FirebaseFirestore | firestore.CollectionReference<T>,
+  relativePath?: string
+): DocumentReference {
+  // We allow omission of 'pathString' but explicitly prohibit passing in both
+  // 'undefined' and 'null'.
+  if (arguments.length === 1) {
+    relativePath = AutoId.newId();
+  }
+
+  if (!relativePath) {
+    throw new FirestoreError(
+      Code.INVALID_ARGUMENT,
+      `Invalid path (${relativePath}). Empty paths are not supported.`
+    );
+  }
+
+  const path = ResourcePath.fromString(relativePath);
+  if (parent instanceof Firestore) {
+    if (!DocumentKey.isDocumentKey(path)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${path}). Path points to a collection.`
+      );
+    }
+    return new DocumentReference(parent, new DocumentKey(path));
+  } else {
+    const coll = hardCast(parent, CollectionReference);
+    const absolutePath = coll._path.child(path);
+    if (!DocumentKey.isDocumentKey(absolutePath)) {
+      throw new FirestoreError(
+        Code.INVALID_ARGUMENT,
+        `Invalid path (${absolutePath}). Path points to a collection.`
+      );
+    }
+    return new DocumentReference(
+      coll.firestore,
+      new DocumentKey(absolutePath),
+      coll._converter
+    );
+  }
+}
+
+export function parent(
+  reference: CollectionReference<unknown>
+): DocumentReference<firestore.DocumentData> | null;
+export function parent<T>(
+  reference: DocumentReference<T>
+): CollectionReference<T>;
+export function parent<T>(
```

Comment:
Combines https://osscs.corp.google.com/firebase-sdk/firebase-js-sdk/+/master:packages/firestore/src/api/database.ts;l=969 and https://osscs.corp.google.com/firebase-sdk/firebase-js-sdk/+/master:packages/firestore/src/api/database.ts;l=969